Implement external IP (LoadBalancer) allocation & announcement via BGP for services

christarazi opened the pull request:
This PR implements external IP allocation and announcement via BGP for
services. It does this by integrating MetalLB into Cilium. For more details,
see the CFP:
https://docs.google.com/document/d/1y47jUlOv2GXsia4vuOIn2yi_sMBB77FvnRheOmlC1a4/edit#heading=h.5mwz58y9syps.

This PR relies on a fork of MetalLB, which has made the code importable from
outside modules (Cilium). See https://github.com/cilium/metallb/pull/2.

The meat of the PR is in the last few commits, namely:
* bgp: Add config package
* bgp: Add central logging package
* bgp: Add k8s package for MetalLB client
* bgp, operator: Implement LB IP allocation via MetalLB
* daemon, bgp, watchers: Implement LB IP announcement via BGP 

All other commits are refactors or add
small utility functionality to ease the later commits.

See commit msgs.

Fixes: https://github.com/cilium/cilium/issues/14395

---

## TODO:

- [x] Get https://github.com/cilium/cilium/pull/15295 merged
- [x] Merge https://github.com/christarazi/metallb/pull/1 and move under `cilium/metallb`
- [x] Finish CI test
- [x] Address TODOs
- [x] ~Finish documentation~ Moved to https://github.com/cilium/cilium/pull/15519

## Followups: 
Moved to https://github.com/cilium/cilium/issues/15611

christarazi committed:
operator/watchers: Separate out service watcher initialization

The service controller inside the Operator is created inside
StartSynchronizingServices(). This function is only called when the
kvstore is configured and `--synchronize-k8s-services` is passed to the
Operator. Besides creating a service controller, it also performs other
tasks unrelated to services, which makes it unusable for reuse.

This commit refactors the service synchronization in the watcher package
to move the controller outside of StartSynchronizingServices(). In
future commits, creating a service controller for BGP announcement of
service IPs will be implemented, and thus depends on this refactor.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
operator: Move leader election vars under the correct section

They were declared under the "Azure" section and this commit moves them
to the general vars section.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon, install, operator, option: Introduce BGP flags

This option is shared between the Agent and the Operator, and will be
required to be passed to both. This commit also updates the Helm charts
with the new flags.

BGP configuration is passed via a file path, which when running with K8s
will point to a mounted ConfigMap created by the user before starting
Cilium.

Due to the lack of upstream support for EndpointSlices in MetalLB, we
must fallback to using the regular Endpoints object.

See https://github.com/metallb/metallb/issues/811.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
bgp, operator: Implement LB IP allocation via MetalLB

This commit implements LoadBalancer IP allocation via MetalLB
integration for service IP announcement via BGP.

This is done by monitoring all service objects via K8s watchers. Each
object seen by the service watcher is pushed into a BGP Manager queue
where they are processed by calling down to the MetalLB reconciliation
logic. If a reconciliation fails (types.SyncStateError), then the object
is re-added into the queue. Once reconciliation for an object returns
types.SyncStateSucess, then it is removed from the queue. During
reconciliation, it's possible for MetalLB to return
types.SyncStateReprocessAll under two circumstances:

1) A service is deleted
2) Configuration changes

(1) is explained below. (2) does not apply to Cilium because our BGP
configuration is static and cannot change dynamically, therefore we
don't need to react to this event.

The MetalLB reconciliation occurs when a service of type LoadBalancer is
found. Once found, It'll allocate an LB IP for it from the configured IP
pool(s), which is provided by the user. Once the IP is allocated, the
Operator will update the service object accordingly. The Agents will
receive this update and announce via BGP (announcement varies depending
on externalTrafficPolicy=Cluster or Local).

The Operator now requires the use of the cache.Store (indexer)
associated with the informer (cache.Controller). Previously, we avoided
the use of the indexer in favor of our own K8s service cache. However,
it is not applicable for this BGP implementation because our service
cache stores services in an internal representation of a service
(k8s.Service). This makes it unusable because the types are obviously
incompatible with MetalLB's API.

In addition, we need to be able to re-list all services in the cluster.
If we had to re-list all the services from the cache and convert their
type, it would be a purely wasteful (useless) operation. Instead,
keeping the services in their original type is a more efficient use of
compute resources. The reason we need to re-list the services anyway is
in the case of the LB IP pool(s) being exhausted. If the pool(s) are
exhausted, then new services will sit in a pending state waiting for
their LB IP to be allocated. Only when a service is deleted
(reconciliation returns types.SyncStateReprocessAll) can we reassign IPs
to pending services. Therefore, we must re-list all the services upon a
service delete event to find which services are in need of an LB IP.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
vendor: Add go.universe.tf/metallb

For now use fork of MetalLB until the changes have been upstreamed.

This bumps the github.com/miekg/dns module because it is a transient
dependency, but we have a replace directive for it anyway.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s, k8s/slim: Add LoadBalancerIP field

This field is crucial for BGP integration with MetalLB because a user
can perfer an LB IP by setting this field, rather than allocate a random
IP for an IP pool.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
install: Grant cilium-operator update permission for services

This is needed in order to update the status of service objects from the
Operator, in order to set the LB IP allocated by MetalLB.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
watchers: Consolidate add and update logic for endpoints

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon, bgp, watchers: Implement LB IP announcement via BGP

This commit implements LoadBalancer IP announcements by integrating MetalLB
natively into Cilium.

Each node in the K8s cluster running a Cilium Agent will be a BGP
speaker. The speaker controller is notified of K8s events, specifically
updates to services, endpoints, and nodes.

The speaker is interested in the following K8s objects and events:

Services  -> OnUpdateService(), OnDeleteService()
Endpoints -> OnUpdateEndpoints()
Nodes     -> OnUpdateNode()

For Services updates, if the object contains a service of type
LoadBalancer with an LB IP (allocated by the Operator, see previous
commit), then the Agents will consider whether to advertise the LB IP to
the configured BGP router (provided by the user), depending on whether
the service's externalTrafficPolicy (either Cluster or Local). For the
difference between the two, see https://metallb.universe.tf/usage/#bgp.
This is also dependant on whether the service has healthy backends
(endpoints) running.

For Service deletions, if the speaker was previously announcing the
service via BGP, then it will withdraw that advertisement. This will
cause the BGP router to remove the route to the node advertising that it
can route the LB IP.

For Endpoints updates, we watch for these because we must know the
health of the backends (endpoints). When all backends of a service are
healthy, only then is a BGP announcement made. If the backends become
unhealthy, then the announcement is withdrawn. A mapping is stored
internally for services to endpoints, so that we may cross-reference
them back to a Service.

For Node updates, we watch for these because MetalLB allows users to
limit BGP peers to certain nodes. See
https://metallb.universe.tf/configuration/#limiting-peers-to-certain-nodes
for more details. For this event, we are only interested in the labels
of a Node.

All of these events described above are enqueued into an internal queue
for processing. The processing of the events off of the queue is done in
a separate goroutine. Having a queue allows us to retry on errors in the
reconciliation logic. The reason it's used is to avoid retrying directly
inside the event handlers of the K8s watcher, which would block other
handlers from running. Instead, the event is re-enqueued and eventually
re-processed again at a later time.

Each reconciliation attempt returns a status, either
types.SyncStateSucess, types.SyncStateError, or
types.SyncStateReprocessAll. Any attempt with types.SyncStateError is
retried. For types.SyncStateReprocessAll, MetalLB returns this only when
the BGP configuration changes. However, this does not apply to Cilium
because our BGP configuration is statically provided via CLI flags, Helm
values file, or via ConfigMap. We do not support dynamically reloading
the configuration at runtime, therefore we don't care for
types.SyncStateReprocessAll.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Provide slim Service -> v1 Service helpers

This is needed for the BGP controller inside the Operator. The reason
they're needed is because the Operator watches slim Service objects,
while the BGP controller is based off MetalLB which uses the raw v1
type. Inside the handlers for BGP, a conversion is necessary before
calling into MetalLB.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
bgp: Add config package

This package will allow both the BGP speaker and the BGP controller to
use the same logic for configuration. It currently does not support
loading configuration dynamically at runtime.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
test: Convert LoadBalancer test to use BGP

We can reuse this test that already validates connectivity to the
external IP of a service of type LoadBalancer (LoadBalancerIP).
Previously, we deployed a custom made controller to allocate external
IPs for the sake of testing. Now that BGP announcements of service IPs
has been implemented, let's convert the test to leverage this.

The newly added "frr.yaml.tmpl" and "bgp-configmap.yaml.tmpl" files are
a Go text/template of a YAML file containing the pod spec for the BGP
router and the ConfigMap, respectively. They are templated because we
want to avoid hardcoding node names, node IPs, and BGP router IPs, which
would otherwise make this test un-runnable outside of our CI using
vagrant.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Add endpoints getter from service cache

This will be used in future commits where the mapping between services
and endpoints is crucial for the BGP speaker.

Specifically, whenever there are service resource updates, we need to
ensure that the MetalLB controller receives the correct and most
up-to-date backends for a given service. Otherwise, the BGP
announcements won't work properly as the MetalLB controller won't be
able to determine if the backends are healthy, and therefore might cause
spurious / missed announcements for the LB IP (of the service).

Signed-off-by: Chris Tarazi <chris@isovalent.com>

brb commented on pkg/k8s/watchers/callbacks/callbacks.go:
```diff
@@ -0,0 +1,59 @@
+// creating K8s watchers. The intent is to allow the K8s watchers to
+// consolidate all the event handler callbacks from various subsystems into one
+// place.
+package callbacks
```
Nit: maybe `consumer` or `subscriber`?

aanm replied:
For historic reasons, anything but `consumer`.

christarazi replied:
What are the historic reasons?

christarazi replied:
Either way, I've renamed it to subscriber, but still interested in the reasons :)

brb commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -132,9 +134,11 @@ func StartSynchronizingServices(shared bool, cfg ServiceSyncConfiguration) {
+
 	// Watch for v1.Service changes and push changes into ServiceCache
 	_, svcController := informer.NewInformer(
 		cache.NewFilteredListWatchFromClient(k8s.WatcherClient().CoreV1().RESTClient(),
```
Would it be possible to create a wrapper for the returned value which internally would do `AddFunc: func(obj interface{}) { for _, f := range callbacks { f.AddFunc(obj) } }`? This would allow us to DRY.

brb commented on daemon/cmd/daemon_main.go:
```diff
@@ -1370,6 +1376,11 @@ func initEnv(cmd *cobra.Command) {
+
+	if option.Config.AnnounceLoadBalancerIP {
+		option.Config.EnableNodePort = true
+		log.Infof("Auto-set BPF NodePort (%q) because LB IP announcements via BGP depend on it.", option.EnableNodePort)
```
It should work with kube-proxy, so the auto-enable is not needed.

christarazi replied:
The reason I had to do this is because this code would not trigger otherwise: https://github.com/christarazi/cilium/blob/52c7b67b8e0030d7b6d29f85f3f1eb6eb673c80a/pkg/k8s/service.go#L442

aanm replied:
Leave a comment in the code for that reason.

christarazi replied:
Done

brb commented on daemon/cmd/daemon_main.go:
```diff
@@ -928,6 +928,12 @@ func init() {
+	flags.Bool(option.AnnounceLoadBalancerIP, false, "Announces service IPs of type LoadBalancer via BGP")
+	option.BindEnv(option.AnnounceLoadBalancerIP)
+
+	flags.String(option.BGPConfig, "", "BGP configuration. It should be in JSON or YAML form. This is compatible with MetalLB's configuration.")
```
CC'ing @seanmwinn for possible UX concerns.

pchaigno replied:
@brb - @seanmwinn doesn't seem to watch for GitHub notifications. You'll have to ping him on Slack if you really want an answer :-)

brb commented on pkg/k8s/slim/k8s/api/core/v1/types.go:
```diff
@@ -566,6 +566,14 @@ type ServiceSpec struct {
+	// the loadBalancerIP when a load balancer is created.
+	// This field will be ignored if the cloud-provider does not support the feature.
+	// +optional
+	LoadBalancerIP string `json:"loadBalancerIP,omitempty" protobuf:"bytes,8,opt,name=loadBalancerIP"`
```
I'm quite surprised that we didn't have this field, as for KPR we were successfully provisioning LoadBalancer type services.

christarazi replied:
This field is used when users themselves want to request a specific LB IP. It is _not entirely_ required, but it is checked by the MetalLB logic.

Commit msg:

>This field is crucial for BGP integration with MetalLB because a user
can perfer an LB IP by setting this field, rather than allocate a random
IP for an IP pool.

brb commented on install/kubernetes/cilium/templates/cilium-operator-clusterrole.yaml:
```diff
@@ -43,6 +43,22 @@ rules:
+  - get
+  - list
+  - watch
+  - update
```
Do we need this update? AFAIK only service/status will be updated?

aanm replied:
This :point_up:

christarazi replied:
Hmm, good call out. The reason I added it is because MetalLB has code to [update](https://github.com/metallb/metallb/blob/124f6beb9165078111de9bd32d811c4e94784492/controller/main.go#L80) the object . However, if you follow that code path, there's nothing that MetalLB actually changes in the service's annotations or spec. So upstream MetalLB doesn't need this permission at all either. Upstream PR: https://github.com/metallb/metallb/pull/823. 

(Another positive for that PR is it removes one call to `reflect.DeepEqual` :) )

christarazi replied:
Updated fork PR to remove it and removed it from cilium-operator clusterrole.

brb commented on go.mod:
```diff
@@ -119,6 +120,8 @@ replace (
 	// Using cilium/netlink until XFRM patches merged upstream
 	github.com/vishvananda/netlink => github.com/cilium/netlink v0.0.0-20210223023818-d826f2a4c934
 
+	go.universe.tf/metallb => github.com/christarazi/metallb v0.1.1-0.20210313015101-1f6227890be7
```
Let's move to github.com/cilium/metallb.

joestringer replied:
^ I didn't look at the rest of this but I noticed this comment and set up the Cilium fork:

https://github.com/cilium/metallb

christarazi replied:
Thanks Joe!

brb reviewed (commented):
I did the first round of the reviews. Looks great, so far a few nits.

rolinh commented on pkg/k8s/watchers/callbacks/service.go:
```diff
@@ -0,0 +1,60 @@
+func NewService() *ServiceList {
+	return &ServiceList{
+		cbs: make([]ServiceHandler, 0),
+	}
```
nit: it's unnecessary to call `make` here.
```suggestion
	return &ServiceList{}
```

rolinh commented on pkg/k8s/watchers/callbacks/service.go:
```diff
@@ -0,0 +1,60 @@
+func (l *ServiceList) Callbacks() []ServiceHandler {
+	c := make([]ServiceHandler, 0)
+	l.RLock()
+	copy(c, l.cbs)
```
Nothing will be copied [if `len(l.cbs) == 0`](https://golang.org/pkg/builtin/#copy).
```suggestion
	l.RLock()
	c := make([]ServiceHandler, len(l.cbs))
	copy(c, l.cbs)
```

christarazi replied:
Thanks! Yes I caught this earlier, and updated https://github.com/cilium/cilium/pull/15295, but didn't rebase this PR on top. I didn't want to push too much to this PR in case other were reviewing at the moment.

rolinh commented on pkg/k8s/watchers/callbacks/raw.go:
```diff
@@ -0,0 +1,56 @@
+func (l *RawList) Callbacks() []cache.ResourceEventHandlerFuncs {
+	c := make([]cache.ResourceEventHandlerFuncs, 0)
+	l.RLock()
+	copy(c, l.cbs)
```
Nothing will be copied [if `len(l.cbs) == 0`](https://golang.org/pkg/builtin/#copy).
```suggestion
	l.RLock()
	c := make([]cache.ResourceEventHandlerFuncs, len(l.cbs))
	copy(c, l.cbs)
```

christarazi replied:
See earlier comment

rolinh commented on pkg/bgp/speaker/log.go:
```diff
@@ -0,0 +1,35 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speaker
```
What's the point of this new file? Can't the logger be used directly?

twpayne replied:
Can this code be shared with `pkg/bgp/controller/log.go`?

christarazi replied:
The point of the file is to keep all the irrelevant logger code away from the main code. @twpayne I've moved the logging wrapper of MetalLB to a shared package

rolinh reviewed (changes requested):
I think I've spotted a bug, PTAL.

kkourt commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -16,9 +16,11 @@ cilium-agent [flags]
       --allow-icmp-frag-needed                               Allow ICMP Fragmentation Needed type packets for purposes like TCP Path MTU. (default true)
       --allow-localhost string                               Policy when to allow local stack to reach local endpoints { auto | always | policy } (default "auto")
       --annotate-k8s-node                                    Annotate Kubernetes node (default true)
+      --announce-lb-ip                                       Announces service IPs of type LoadBalancer via BGP
```
Would it make sense to name this `--bgp-announce-lb-ip`? I think it's useful if related options are grouped together in the docs.

christarazi replied:
I like that idea, done.

kkourt commented on pkg/k8s/factory_functions.go:
```diff
@@ -278,6 +278,25 @@ func convertToK8sServicePorts(ports []v1.ServicePort) []slim_corev1.ServicePort
+	}
+
+	ports := make([]v1.ServicePort, 0, len(slimPorts))
+	for _, port := range ports {
```
Is this supposed to be:
```
for _, port := range slimPorts {
```

Or am I missing something?

aanm replied:
Why aren't the unit-tests catching this? :-)

aanm replied:
There are converts all over the place in the kubernetes/kubernetes source code. Might be worth seeing if we can have something similar and if they would be autogenerated.

christarazi replied:
Good catch, thanks! I should've have written unit tests from the beginning :sweat_smile:. Added tests and fixed issue

For autogenerated conversion functions, there's https://pkg.go.dev/k8s.io/code-generator/cmd/conversion-gen which k/k uses. Given that we'd need to setup new tooling for this, I think we can tackle this as a followup?

kkourt commented on pkg/k8s/endpoints.go:
```diff
@@ -287,7 +287,9 @@ func newExternalEndpoints() externalEndpoints {
-	return version.Capabilities().EndpointSlice && option.Config.K8sEnableK8sEndpointSlice
+	return version.Capabilities().EndpointSlice &&
+		option.Config.K8sEnableK8sEndpointSlice &&
+		!option.Config.AnnounceLoadBalancerIP
```
I think a comment here that this is a MetalLB thing and not an `AnounceLoadBalancerIP` thing in general would be helpful.

kkourt commented on operator/watchers/bgp.go:
```diff
@@ -0,0 +1,215 @@
+		if err != nil {
+			return types.SyncStateError
+		}
+		if !exists {
```
Comment above says:
> If the service exists in the service indexer, then it will begin reconciliation.

But the name of the variable seems to seems to suggest that reconciliation happens if service does not exist.

christarazi replied:
Clarified that if the object doesn't exist, then it means that MetalLB should de-allocate the IP

kkourt commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -0,0 +1,217 @@
+			out.NotReady = append(out.NotReady, ep)
+		}
+	}
+	fmt.Printf("chris convertInternalEndpoints ready=%v not-ready=%v\n",
```
:-)

kkourt commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -0,0 +1,217 @@
+			out.Ready = append(out.Ready, ep)
+		}
+		if !be.Ready {
+			out.NotReady = append(out.NotReady, ep)
```
One of my questions while reading the commit message was: will the BGP announcement be withdrawn if the status of a  backend changes from healthy to unhealthy? Am i correct to assume that this happens here?

christarazi replied:
Yes, it will. I thought I mentioned that in the commit msg, but apparently not:

> For Endpoints updates, we watch for these because we must know when a
> service has healthy backends (endpoints). Only when that happens can a
> BGP announcement be made. A mapping is stored internally for services to
> endpoints, so that we may cross-reference them back to a Service. For
> clarity, an announcement is never made unless a service has healthy
> backends.

I will add it in that parapraph and clarify it a bit, because re-reading it now, it doesn't flow as nicely as I intended :). Thanks!

kkourt reviewed (commented):
Thanks for splitting up the PR into reviewable chunks, and the well-written commit messages!

twpayne commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -132,21 +141,69 @@ func StartSynchronizingServices(shared bool, cfg ServiceSyncConfiguration) {
+		endpointController, endpointSliceEnabled = endpointSlicesInit(k8s.WatcherClient(), swgEps)
+		// the cluster has endpoint slices so we should not check for v1.Endpoints
+		if endpointSliceEnabled {
+			break
```
If `endpointSliceEnabled` is `true` then `go endpointController.Run(wait.NeverStop)` (line 167) is not called. For this to be correct, it would mean that looks `endpointSlicesInit` calls `go endpointController.Run` itself whereas it is the caller's responsibility to call `go endpointController.Run` in the case of `endpointsInit`. Is this intentional?

christarazi replied:
Ah, good catch! `go endpointController.Run` should happen after the switch statement, and it should be removed from `endpointSlicesInit()`

Edit: it looks like it's actually intentional. It seems the reason is because when `endpointSliceEnabled` is `true`, then we need to be able to run the controller to sync with K8s to actually determine if endpointslices are truely supported and enabled. I'll add a clarifying comment

twpayne commented on operator/watchers/service_cache.go:
```diff
@@ -0,0 +1,48 @@
+	log.WithField(logfields.ServiceName, obj.Name).Debugf("Received service addition %+v", obj)
+	K8sSvcCache.UpdateService(obj, c.swgSvcs)
+}
+func (c *serviceCacheCallback) OnUpdate(oldObj, newObj *slim_corev1.Service) {
```
nit: missing newlines between functions, also four lines below.

twpayne commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -0,0 +1,217 @@
+	for i := range in.Status.LoadBalancer.Ingress {
+		ing[i].IP = in.Status.LoadBalancer.Ingress[i].IP
+	}
+	fmt.Printf("chris convertService(%v) ingress=%v\n", in.GetName(), ing)
```
nit: remove debug code (or convert into proper calls to `log`).

twpayne commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -0,0 +1,217 @@
+		}
+	}
+	fmt.Printf("chris convertEndpoints(%v) ready=%v not-ready=%v\n",
+		in.GetName(), len(out.Ready), len(out.NotReady))
```
nit: remove debug code (or convert into proper calls to `log`).

twpayne commented on pkg/k8s/watchers/endpoint.go:
```diff
@@ -88,12 +90,15 @@ func (k *K8sWatcher) endpointsInit(k8sClient kubernetes.Interface, swgEps *lock.
 func (k *K8sWatcher) updateK8sEndpointV1(oldEP, newEP *slim_corev1.Endpoints, swg *lock.StoppableWaitGroup) error {
 	k.K8sSvcCache.UpdateEndpoints(newEP, swg)
+	if option.Config.AnnounceLoadBalancerIP {
+		fmt.Printf("chris updateK8sEndpointV1 %v\n", newEP.GetName())
```
nit: remove debug code (or convert into proper calls to `log`).

twpayne commented on pkg/k8s/watchers/service.go:
```diff
@@ -89,6 +91,10 @@ func (k *K8sWatcher) addK8sServiceV1(svc *slim_corev1.Service, swg *lock.Stoppab
 		}
 	}
+	if option.Config.AnnounceLoadBalancerIP {
+		fmt.Printf("chris addK8sServiceV1 %v\n", svc.GetName())
```
nit: remove debug code (or convert into proper calls to `log`).

twpayne commented on pkg/k8s/watchers/service.go:
```diff
@@ -104,5 +110,9 @@ func (k *K8sWatcher) deleteK8sServiceV1(svc *slim_corev1.Service, swg *lock.Stop
 		}
 	}
+	if option.Config.AnnounceLoadBalancerIP {
+		fmt.Printf("chris deleteK8sServiceV1 %v\n", svc.GetName())
```
nit: remove debug code (or convert into proper calls to `log`).

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -928,6 +928,12 @@ func init() {
+	flags.Bool(option.AnnounceLoadBalancerIP, false, "Announces service IPs of type LoadBalancer via BGP")
+	option.BindEnv(option.AnnounceLoadBalancerIP)
+
+	flags.String(option.BGPConfig, "", "BGP configuration. It should be in JSON or YAML form. This is compatible with MetalLB's configuration.")
```
as a user I'm not sure if I understand this flag. If this flag is compatible with metalb's why not make it explicit as such? something like `bgp-config-metallb`?

joestringer replied:
If this is really a full JSON/YAML blob, can we handle it through a filepath (optionally mounted from some configmap resource)? Passing a giant configuration blob into an argument doesn't seem super user-friendly.

EDIT: Oh, Andre mentioned this below also :sweat_smile:

christarazi replied:
Updated the code to use a dedicated configmap instead

aanm commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -680,6 +680,14 @@ data:
 
+{{- with .Values.bgp.config }}
+  bgp-config: |
+{{ toYaml . | indent 4 }}
```
oh, this is the actual yaml?! :exploding_head: I would suggest to make the `bgp-config` defined a a configmap in k8s and mount it as a file in the cilium-container. Thus, this `bgp-config` would be the filepath.

christarazi replied:
Done

aanm commented on install/kubernetes/cilium/templates/cilium-operator-clusterrole.yaml:
```diff
@@ -43,6 +43,22 @@ rules:
+- apiGroups:
+  - ""
+  resources:
+  # to perform LoadBalaner IP allocation for BGP
```
typo LoadBalancer

aanm commented on operator/option/config.go:
```diff
@@ -17,6 +17,8 @@ package option
 import (
 	"time"
 
+	opt "github.com/cilium/cilium/pkg/option"
```
I think we were trying to avoid importing `pkg/option` into the operator/option? cc @tklauser

tklauser replied:
I think importing `pkg/option` from `operator` is currently OK (we do it in other places).

Eventually I'd like to split out agent-specific flags out of `pkg/option` so agent/operator specific flags live in their own package and `pkg/option` only contains generic shared flags.

christarazi replied:
I've duplicated the options anyway so that we don't need to import all the agent flags into the operator. Agree on the long-term goal :+1:.

aanm commented on operator/watchers/bgp.go:
```diff
@@ -0,0 +1,215 @@
+// Copyright 2021 Authors of Cilium
```
Might be wroth putting this in its own package in case we want to create an operator image without BGP, for example a cloud provider having its own BGP integration, then we wouldn't need to compile this in the operator.

christarazi replied:
In general, I'm all for smaller binary sizes, but BGP actually doesn't increase the Operator's binary size by much at all (~200KB):

Before:
```
.rwxr-xr-x  67,248,128 chris 25 Mar 18:51  cilium-operator
.rwxr-xr-x  64,610,304 chris 25 Mar 18:51  cilium-operator-aws
.rwxr-xr-x  50,225,152 chris 25 Mar 18:51  cilium-operator-azure
.rwxr-xr-x  47,628,288 chris 25 Mar 18:51  cilium-operator-generic
```

After:

```
.rwxr-xr-x  67,489,792 chris 25 Mar 18:52  cilium-operator
.rwxr-xr-x  64,847,872 chris 25 Mar 18:52  cilium-operator-aws
.rwxr-xr-x  50,470,912 chris 25 Mar 18:52  cilium-operator-azure
.rwxr-xr-x  47,869,952 chris 25 Mar 18:52  cilium-operator-generic
```

Do you think it's worth all the effort to provide it as a separate binary? The user can simply not enable BGP (`--bgp-announce-lb-ip=false`) and there would be no impact.

aanm replied:
just mostly because `operator/watchers` doesn't need to have any code specific for `bgp`, it should be at least in its own package.

aanm commented on operator/watchers/bgp.go:
```diff
@@ -0,0 +1,215 @@
+	}
+}
+
+func toV1Service(in *slim_corev1.Service) *v1.Service {
```
How will we be sure new fields added in v1 Spec won't be forgotten to be added here?

christarazi replied:
Good question, I think this can be solved by using conversion-gen as mentioned in a previous comment.

aanm commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -147,6 +147,15 @@ type redirectPolicyManager interface {
 
+type bgpSpeakerManager interface {
+	OnUpdateService(name string, svc *slim_corev1.Service)
+	OnDeleteService(name string, svc *slim_corev1.Service)
```
Why is the name being passed on? This is already part of the `svc` that is also being passed on. Also, the `name` is not unique since we can have the same name in multiple namespaces.

christarazi replied:
Good point, I changed this to use `k8s.ServiceID` instead which already parses the ns + name and provides a `String()` that we can use to pass down to MetalLB.

aanm commented on pkg/bgp/speaker/events.go:
```diff
@@ -0,0 +1,78 @@
+
+	switch k := key.(type) {
+	case svcEvent:
+		e := key.(svcEvent)
```
there's no need to do this, line `64` already has the converted type, so in here `k` is already `svcEvent`. Same applies for line 69 and 72

christarazi replied:
You're right. For some reason, I thought I would get compilation errors if I didn't, but turns out there's no need.

aanm commented on pkg/bgp/speaker/events.go:
```diff
@@ -0,0 +1,78 @@
+		st := s.do(key)
+		switch st {
+		case types.SyncStateError:
+			s.queue.AddRateLimited(key)
```
Do we want to have rate limiting in service events? What will happen if we receive 1k service events and this watcher can't keep up with all events? a) Does it mean all other watchers will be blocked by this watcher? b) Or we have an independent queue that will store all svc events received and process them in parallel? If it is b) what guarantees do we have that we won't have concurrent issues while handling events for a svc that no longer exists, or it was deleted and re-created with the same name?

christarazi replied:
Good point. The way it's implemented is (b). The queue processing runs in a goroutine and the handler pushes into the queue, which is not blocking. So other handlers should not be affected. However, you also make a good 2nd point about concurrent issues. Rather than thinking about a more complex way to handle them, let's use a simple non-rate limiting queue. Plus, MetalLB isn't performing any complex operations; it's simply allocating an IP. Failing to do that is the only real consequential failure. I don't suspect that to happen often, besides the case where the user exhausts their IP pool . Therefore, we don't really need to rate-limit the retries.

aanm commented on operator/watchers/service_cache.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+func (c *serviceCacheCallback) OnAdd(obj *slim_corev1.Service) {
+	log.WithField(logfields.ServiceName, obj.Name).Debugf("Received service addition %+v", obj)
```
I'm not sure if it is worth duping the entire object into debug messages. I know it must have been useful while developing but it doesn't seem to be useful now the functionality is implemented. Consider duping only key-fields that matter to debug an issue.

christarazi replied:
I copied these log statements straight from the service watcher callbacks, so we were already doing this before. I was just trying to keep the code the same.

aanm commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -132,21 +141,69 @@ func StartSynchronizingServices(shared bool, cfg ServiceSyncConfiguration) {
-					K8sSvcCache.UpdateService(k8sSvc, swgSvcs)
+					for _, f := range serviceCallbacks.Callbacks() {
+						f.OnAdd(k8sSvc)
+					}
```
To avoid doing a copy of callbacks every time we receive an event consider executing a function in `serviceCallbacks` that iterates over all callbacks. Something like a). Or consider `panic` if the developer registers another callback after the first callback function call. We do a similar thing like this in other places @brb might help since he is an author of some of that code as I can't remember where it is at this moment.

a)
```golang
func (l *ServiceList) AddAllCallbacks(svc *slim_corev1.Service) {
	for _, cb := range l.cbs {
		cb.OnAdd(svc)
	}
}
```
```golang
				if k8sSvc := k8s.ObjToV1Services(obj); k8sSvc != nil {
					serviceCallbacks.OnAdd(k8sSvc)
				}
```

brb replied:
@aanm Which code?

christarazi replied:
Implemented ability to notify call subscribers without creating copies.

aanm reviewed (changes requested):
Great work @christarazi!

I've left some comments overall. Let me know if you have any questions about my feedback over slack.

christarazi committed:
bgp: Add central logging package

This package will be shared by the controller and speaker BGP packages
that will be added in future commits.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
operator/watchers: Clarify logic inside endpoint controller init

Signed-off-by: Chris Tarazi <chris@isovalent.com>

nathanjsweet commented on Documentation/gettingstarted/bgp.rst:
```diff
@@ -0,0 +1,149 @@
+       - name: default
+         protocol: bgp
+         addresses:
+         - 192.168.10.0/24
```
Here and in other places I'm finding the usage of a known private address range somewhat confusing. Aren't these typically public IP addresses? If so, I think we should use a public range as an example.

christarazi replied:
Good point, will change

joestringer replied:
^ Let's be careful here, I don't think we want to be creating examples that may accidentally advertise real ranges that other organizations are responsible for. Maybe something like a `192.0.*` range from [RFC5735](https://tools.ietf.org/html/rfc5735)?

christarazi replied:
Done

nathanjsweet commented on Documentation/gettingstarted/bgp.rst:
```diff
@@ -0,0 +1,149 @@
+    $ kubectl get svc
+    NAME        TYPE          CLUSTER-IP  EXTERNAL-IP    PORT(S)       AGE
+    kubernetes  ClusterIP     172.20.0.1  <none>         443/TCP       4d23h
+    test-lb     LoadBalancer  172.20.0.5  192.168.10.10  80:30724/TCP  10s
```
See my other comment about using a private IP address as a public example.

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -0,0 +1,216 @@
+		DisableLayer2: true,
+	})
+	if err != nil {
+		panic(err)
```
Replace panic with fatal

anfernee commented on operator/option/config.go:
```diff
@@ -353,6 +358,10 @@ func (c *OperatorConfig) Populate() {
 	c.LeaderElectionRetryPeriod = viper.GetDuration(LeaderElectionRetryPeriod)
+	c.BGPAnnounceLBIP = viper.GetBool(opt.BGPAnnounceLBIP)
+	if c.BGPAnnounceLBIP {
+		c.SyncK8sServices = true
```
Consider log the overwriting to avoid confusion, like previously in agent?

christarazi replied:
Good idea, done.

aanm commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -547,6 +552,12 @@ spec:
+{{- if .Values.bgp.lbIP }}
+      - configMap:
+          name: bgp-config
+          optional: true
```
This is a tricky one, it we could consider removing this `optional` since the user has explicitly set `bgp.lbIP` in their helm configuration.

christarazi replied:
Fixed

aanm commented on install/kubernetes/cilium/templates/cilium-operator-deployment.yaml:
```diff
@@ -278,4 +283,10 @@ spec:
+{{- if .Values.bgp.lbIP }}
+      - configMap:
+          name: bgp-config
+          optional: true
```
same here

aanm commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1580,3 +1580,8 @@ clustermesh:
+# -- Configure BGP
+bgp:
+  # -- Enable announcement of service LoadBalancer IPs
+  lbIP: false
```
similar to other options like `externalWorkloads`, shouldn't we have a specific flag named `enabled`?

christarazi replied:
It's a bit tricky with BGP because `enabled` alone doesn't mean anything. We need to know if the user is interested in LB IP announcements or PodCIDR announcements (coming soon), etc. Hence why there's `lbIP`. Any ideas?

aanm replied:
@christarazi what about:
```
bgp:
   enabled: true|false
   announce:
      lbIP: true|false
      podCIDR: true|false
```

christarazi replied:
But what does the following mean?

```
bgp:
   enabled: true
   announce:
      lbIP: false
      podCIDR: false
```

aanm replied:
That's a user mistake, we can detect that in helm and warn the user about that.

christarazi replied:
Fair enough, will implement.

christarazi replied:
Done

aanm commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -0,0 +1,221 @@
+		// }
+		// if !be.Ready {
+		// 	out.NotReady = append(out.NotReady, ep)
+		// }
```
To be ready or not to be ready. That is the question or is the question why is this commented?

christarazi replied:
I had squashed a change with those commented out, with the intent that I don't need the code anymore. But I forgot to remove the commented out code :sweat_smile:

aanm commented on pkg/k8s/watchers/subscriber/raw.go:
```diff
@@ -0,0 +1,70 @@
+		s.OnDelete(obj)
+	}
+	l.RUnlock()
+}
```
use `defer l.RUnlock()` here. If any of the callers of the `s.On*` functions panic, the lock will never be unlocked unless we use a `defer`

christarazi replied:
Done, see the change on the original PR: https://github.com/cilium/cilium/pull/15295

aanm commented on pkg/k8s/watchers/subscriber/service.go:
```diff
@@ -0,0 +1,75 @@
+		s.OnDelete(svc)
+	}
+	l.RUnlock()
+}
```
same here for the `defer`

kkourt commented on pkg/bgp/controller/controller.go:
```diff
@@ -0,0 +1,68 @@
+	f, err := os.Open(option.Config.BGPConfigPath)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to open BGP config file")
+	}
```
`defer f.Close()`?

christarazi replied:
Thanks, good catch

kkourt commented on pkg/bgp/controller/subscriber.go:
```diff
@@ -0,0 +1,205 @@
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		s.queue.Add(svcEvent(key))
+	}
```
Should we log the errors here (and in the subsequent functions)?

christarazi replied:
I've added simple warnings similar to what we do inside `pkg/k8s/factory_functions.go` when a K8s object is not how we expect it. 

In terms of adding more logging throughout, MetalLB code already logs events internally. However I've decided to integrate those logs into Cilium only when debug is enabled, because they can be very chatty. I think as a followup, we should consider more deeply how one might introspect the state of BGP in Cilium without having to turn on debugging & look at chatty logs. I've added that as a followup item because MetalLB itself doesn't provide much itself (https://metallb.universe.tf/configuration/troubleshooting/), so we'd have to implement this ability ourselves.

kkourt reviewed (approved):
Other than what seems to be a  forgotten `Close()` and a minor comment, LGTM. Thank you for the detailed and well-written commit messages!

brb commented on pkg/bgp/controller/subscriber.go:
```diff
@@ -0,0 +1,217 @@
+// Copyright 2021 Authors of Cilium
```
Should we add MetalLB authors to this module authors? I guess lots of code is taken from MetalLB?

christarazi replied:
Done

brb reviewed (changes requested):
:rocket: 

Let's move the forked metallb dependency to https://github.com/cilium/metallb

brb reviewed (approved):
:rocket: :rocket: :rocket:

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1581,3 +1581,12 @@ clustermesh:
+  enabled: false
+  announce:
+    # -- Enable allocation and announcement of service LoadBalancer IPs
+    lbIP: false
```
nit, `lbIP` is rather terse. Any reason to avoid `loadbalancers` / `loadbalancerIPs` / `loadbalancerServices`?

christarazi replied:
Went with `loadbalancerIP`

joestringer commented on install/kubernetes/cilium/templates/cilium-operator-clusterrole.yaml:
```diff
@@ -43,6 +43,21 @@ rules:
+- apiGroups:
+  - ""
+  resources:
+  # to perform LB IP allocation for BGP
```
nit, this comment is above what looks to be the read operations of get, list, watch rather than the write operation of `update` for status below.

joestringer commented on go.mod:
```diff
@@ -56,7 +56,7 @@ require (
 	github.com/kr/pretty v0.2.1
 	github.com/mattn/go-shellwords v1.0.10
-	github.com/miekg/dns v1.0.14
+	github.com/miekg/dns v1.1.26
```
We still `replace` miekg/dns below, does this line have any effect?

joestringer replied:
^ Probably just an automatic fetch as part of updating dependencies, we can ignore this.

joestringer commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -141,10 +148,63 @@ func StartSynchronizingServices(shared bool, cfg ServiceSyncConfiguration) {
+	// This ensures we only have one instance of a service controller.
+	if serviceController != nil {
+		return
+	}
```
:thinking: This is not threadsafe is it? If we really want to protect this, we would usually use `sync.Once`.

christarazi replied:
Good idea

joestringer commented on pkg/k8s/endpoints.go:
```diff
@@ -357,7 +357,11 @@ func newExternalEndpoints() externalEndpoints {
+	// https://github.com/metallb/metallb/issues/811.
+	return version.Capabilities().EndpointSlice &&
+		option.Config.K8sEnableK8sEndpointSlice &&
+		!option.Config.BGPAnnounceLBIP
```
Is this enough? If k8s is configured to use endpoint slices, then won't it disable the logic to handle regular endpoint resources?

Basically I'm wondering if we have to make this more explicit up front (probably just via docs, and/or maybe detecting the incongruance between configurations and warn the user).

christarazi replied:
This should be enough. EndpointSlices don't replace Endpoints, but rather are an alternative, if available. In other words, if the apiserver supports them, but Cilium does **not** use them, then the Endpoint resource is still available for use. And that's what this code snippet is doing. 

However, I still think it's a good idea to add a warning as well, in case the user misses this call out in the docs. (See limitation section https://github.com/cilium/cilium/pull/15519/files#diff-9c888dd144df5108aaa6e5594733faf9522a81d7b6622c39c1cb973ff4a1644aR144)

Btw, I decided to drop this commit entirely actually and and move to disabling logic to simply toggling `option.Config.K8sEnableK8sEndpointSlice` while also logging the warning at the same time. That renders this commit useless, hence dropped.

joestringer commented on pkg/bgp/speaker/log.go:
```diff
@@ -0,0 +1,22 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-spkr")
```
nit: 

```suggestion
var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-speaker")
```

(In general we tend to follow the path under `pkg/` and replace `/` with `-` to determine the subsystem)

christarazi replied:
Thanks, was not aware we had an existing naming scheme for this

joestringer commented on pkg/k8s/watchers/endpoint.go:
```diff
@@ -88,12 +88,14 @@ func (k *K8sWatcher) endpointsInit(k8sClient kubernetes.Interface, swgEps *lock.
 func (k *K8sWatcher) updateK8sEndpointV1(oldEP, newEP *slim_corev1.Endpoints, swg *lock.StoppableWaitGroup) error {
 	k.K8sSvcCache.UpdateEndpoints(newEP, swg)
+	if option.Config.BGPAnnounceLBIP {
+		k.bgpSpeakerManager.OnUpdateEndpoints(newEP)
```
I note that we ignore `oldEP` here. Seems like the BGP library similarly only has a "set" and does not attempt any diff operations based on the transition from old->new endpoints. I didn't yet see the tests associated with the PR, but to make sure that this performs the way we'd expect, this is typically the sort of area where some integration tests could help to ensure that when services are added/deleted over time (causing updates, not just straight up add/delete), the BGP announcements stay in sync with the resources in k8s.

christarazi replied:
I think that's a good idea. I'll take that as a followup to add more tests. I initially had some difficulty in getting the fake k8s client working to do this and decided to move on to get this PR out.

joestringer commented on pkg/bgp/log/log.go:
```diff
@@ -0,0 +1,53 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package log provides a central logging package that wraps MetalLB's logging.
```
nit in the commit message: bpg->bgp.

joestringer commented on pkg/bgp/controller/k8s.go:
```diff
@@ -0,0 +1,45 @@
+	return c.CoreV1().Services(svc.GetNamespace()).Update(context.TODO(), svc, metav1.UpdateOptions{})
+}
+func (c *client) UpdateStatus(svc *v1.Service) error {
+	_, err := c.CoreV1().Services(svc.GetNamespace()).UpdateStatus(context.TODO(), svc, metav1.UpdateOptions{})
```
These `context.TODO()` concern me a bit. Does this mean that if there's some k8s connectivity issue, BGP logic may freeze up without visibility into what's going on?

christarazi replied:
Yeah, ideally we should not be using `context.TODO()` anywhere in Cilium, and especially while in communication with the apiserver. We have an open issue to track this (https://github.com/cilium/cilium/issues/10655), but haven't had the time to go through it. (On me as I'm assigned to it :sweat_smile:)

FWIW, this hasn't caused us problems in the past, but that doesn't preclude it from happening in the future. I'll take a note to followup on this, but this time specifically focus on the BGP subsystem instead of tackling the aforementioned issue as a whole. The reason it's a followup is because I don't think it's as simple as just slapping on a 2m timeout (for ex); we will need to think about the error flow and what actually happens when the context deadlines, etc.

joestringer replied:
> this hasn't caused us problems in the past

[citation needed] ;)

Main reason I bring this up is that we had a whole bunch of trouble with the etcd clients when testing high availability / failover scenarios and it took quite a bit of work to get those codepaths into a decent state. Beyond that, if we just keep adding `context.TODO()` into the codebase then we're going backwards on progress to #10655.

I would consider this a requirement before marking this feature as stable but I agree it needs more careful thought & consideration.

joestringer commented on pkg/bgp/controller/controller.go:
```diff
@@ -0,0 +1,70 @@
+// MetalLB logic for allocate LB IP addresses for service objects of type
+// LoadBalancer.
+type Controller struct {
+	*metallbctl.Controller
```
nit: In Cilium, "controller" has specific meaning around `pkg/controller` which implies an inner loop of logic, status visibility in `cilium status`, etc. This does not appear to be a Cilium controller. At the same time, I don't think there's a notion of a BGP controller (at least, I've not heard such a term outside of SDN-based approaches to BGP where they inherit the term from the SDN side rather than the BGP side).

Is there another term we could use here? For lack of a better options, I think we've tended to use "Manager" as the generic catchall for this type of intermediate object in Cilium code where it's a dedicated module with inputs and outputs and some core logic for handling the relationship between them.

joestringer replied:
^ I think a certain degree of the above discussion is that I don't understand the distinction between the controller vs. subscriber aspects of this package. Clearly something needs to subscribe to k8s events, process them, and emit events into MetalLB to convert into the BGP equivalent.  That object needs to implement `pkg/k8s/watchers/subscriber.ServiceHandler` to handle the service updates, and it needs to initialize & interact with the underlying `metallbctl.Controller` which has the core BGP logic. 

In the current logic though, the calling code in the operator just does `controller.New()` and then directly passes that into `controller.NewSubscriber()`, which returns the ServiceHandler implementation that it can register into the Service updates logic.

Why does it help to separate this logic into two separate pieces which the callers must know about and put together?

joestringer replied:
This also may partially depend on how we plan to extend the logic in future, so depending on how well we understand the direction we're going, it might not make sense to make changes to the structure here for now.

christarazi replied:
I've renamed it to `Manager` because I agree that we shouldn't co-opt the `Controller` notion of Cilium and ultimately cause confusion. 

Good point on the subscriber vs "manager". They actually don't make sense as separate pieces and really should be just one object. I think it turned out this way as an artifact of how I started, but I never revisited the structure as the code evolved. Thanks for the close inspection. :rocket:

joestringer commented on pkg/bgp/controller/log.go:
```diff
@@ -0,0 +1,22 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-ctrl")
```
```suggestion
var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-controller")
```

(Or I guess `manager` if we think it's worth renaming)

joestringer commented on pkg/bgp/speaker/k8s.go:
```diff
@@ -0,0 +1,45 @@
+}
+func (*client) Errorf(_ *v1.Service, desc, fmt string, args ...interface{}) {
+	log.WithField("event", desc).Errorf(fmt, args...)
+}
```
Why is this code copy+pasted with a similar implementation in the other package? If I follow correctly, different binaries are invoking the different paths but I don't see why that should require multiple copies of some of the shared underlying implementations.

(It's only a small bit of code so not such a big deal, but it also suffers from the same `context` limitation)

christarazi replied:
I initially kept it as duplicated code because it was such a small amount of code, but then again, as we expand on BGP, this package might evolve, so it makes sense to refactor this to a common place for future-proofing.

joestringer commented on pkg/k8s/watchers/node.go:
```diff
@@ -35,6 +36,13 @@ func (k *K8sWatcher) nodesInit(k8sClient kubernetes.Interface) {
 		&slim_corev1.Node{},
 		0,
 		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if node := k8s.ObjToV1Node(obj); node != nil {
+					if option.Config.BGPAnnounceLBIP {
+						k.bgpSpeakerManager.OnUpdateNode(node) // Need to seed the BGP speaker
+					}
+				}
+			},
```
:thinking: why did we never subscribe to these events before? Seems a bit odd.

christarazi replied:
Good question, :shrug:

joestringer replied:
cc @aanm do you happen to know why we didn't previously care about Node Add events?

aanm replied:
I'll defer this to @pchaigno since this watcher was only used labels in the host endpoint. @pchaigno How is the host endpoint getting the labels from a k8s node when the agent starts?

joestringer replied:
Oh I see, that makes sense. We wouldn't ever see an Add event for the local node because it would have to be provisioned before Cilium can run on it.

joestringer replied:
^^ @christarazi does that mean we don't actually need this snippet? Seems like we're only listening for node updates for the local node (`fields.ParseSelectorOrDie("metadata.name="+nodeTypes.GetName())`)

pchaigno replied:
> @pchaigno How is the host endpoint getting the labels from a k8s node when the agent starts?

My understanding was the same as Joe's: ADDs can't happen because we're only watching for updates on the local node.

christarazi replied:
@joestringer This snippet is needed to "seed" the MetalLB logic with the node labels which are needed to decide whether a node should be considered for peering. Essentially, this allows you to designate which nodes should connect to the BGP router(s) and ultimately share routes to. I've found that when the K8s node watcher is initialized, it'll trigger with an add event, and not with an update event which the BGP code is already hooked up with. Basically, what hooking into an add event gives us is a signal of when the node labels are available to seed MetalLB with. This is probably not the only way to do this, but I'm not sure of another part of Cilium where we care for the node labels in such fashion (maybe during init?). However, hooking into the K8s watcher seemed like a natural place to do this especially because BGP support depends on the K8s watcher already anyway.

joestringer replied:
> I've found that when the K8s node watcher is initialized, it'll trigger with an add event, and not with an update event

That's a super interesting observation and suggests that we should be handling this event for the hostfw case as well.

At a high level my question is "Is there a bug here that we should prepare a fix for and backport that?".

christarazi replied:
Stepping back a bit, I think we do have a bug. It seems that we mis-assumed that it was impossible to receive a node add event for the local node. That actually doesn't make much sense if we think about it. The K8s watcher is essentially a cache of the K8s resources in Cilium. When Cilium is coming up, it'll initialize this cache. When the informers (watchers) first come up, they'll sync up with the apiserver and populate the local cache. During this, each new resource seen will  trigger an add event for each resource, and in this case, an add event for the local node.

In the moment, I didn't really realize this assumption before. I figured we simply didn't care for node add events.

cc @aanm to double check me on this (it at least fits with my observations)

aanm replied:
> During this, each new resource seen will trigger an add event for each resource, and in this case, an add event for the local node.

correct.

pchaigno replied:
> > I've found that when the K8s node watcher is initialized, it'll trigger with an add event, and not with an update event
> 
> That's a super interesting observation and suggests that we should be handling this event for the hostfw case as well.
> 
> At a high level my question is "Is there a bug here that we should prepare a fix for and backport that?".


It looks like this could be the cause for bugs https://github.com/cilium/cilium/issues/13676 and/or https://github.com/cilium/cilium/issues/13455. I'll send a pull request, hopefully today.

joestringer commented on pkg/k8s/watchers/service.go:
```diff
@@ -104,5 +107,8 @@ func (k *K8sWatcher) deleteK8sServiceV1(svc *slim_corev1.Service, swg *lock.Stop
 			k.redirectPolicyManager.OnDeleteService(svcID)
 		}
 	}
+	if option.Config.BGPAnnounceLBIP {
+		k.bgpSpeakerManager.OnDeleteService(svc)
```
I'd much prefer if all of the dedicated BGP handling logic was under one package. In general what that would mean is during initialization, any other logic that wants to learn about various events should subscribe into the watcher, then in the watcher we would just iterate through all of the generic subscribers that need the service updates, no option config checks here.

Given that we're already leaking such details into the k8s watchers and it's just a couple of cases so far I'm inclined to let this slide in this case but we should probably start thinking about how to keep the core logic free of all of these extra conditionals and references to various features.

joestringer replied:
^ Looks like you already raised this concern in the PR description as a concern. :+1:

christarazi replied:
Yes, this will be addressed in https://github.com/cilium/cilium/issues/15471.

joestringer reviewed (commented):
Nice work! Most of my comments below are minor nits. 

My main concerns would be these:
* Introduction of new `context.TODO()` of which I don't fully understand the repercussions but they seem like they could cause the BGP logic to arbitrarily block/stop working and I don't think there's any user visibility into this
* Introspection. The typical metrics in Cilium are logs, metrics, and status APIs. To some degree, I think the MetalLB structure is working against us here.
* Documenting this well. Intentions around alpha/beta. How we communicate the configuration format since we plan to rely on MetalLB. How scale limitations may impact the implementation in future. Interactions with other options like EndpointSlices (+k8s configuration); reliance on stack routing (therefore the new BPF host routing should be disabled).

I think that in terms of getting something working, it seems like we're practically there. In terms of supporting it or helping users when they hit problems, I think there's a little ways left to go. I think if we're communicating well that this is a fairly early feature and subject to change, then I'm OK with merging more-or-less as-is. Please take a look at the feedback and consider whether you think any of the items should be addressed now before merge or which you think make more sense as follow-up items.

joestringer commented on pkg/bgp/speaker/events.go:
```diff
@@ -0,0 +1,77 @@
+// cannot be keys in a map, which queue is.
+type nodeEvent *map[string]string
+
+// run runs the reconciliation loop, fetching events off of the queue to
+// process. The events supported are svcEvent, epEvent, and nodeEvent. This
+// loop is only stopped (implicitly) when the Agent is shutting down.
+//
+// Adapted from go.universe.tf/metallb/pkg/k8s/k8s.go.
+func (s *Speaker) run() {
```
Oh, and I think there's commonality here between this logic, the core of `pkg/serializer`, and the core of `pkg/eventqueue` where I think we have the opportunity to combine & improve one implementation of a serialized queue that handles concurrent queueing, provides introspection capabilities, etc. I don't think it's reasonable to make any such change right now but it might be worth familiarizing yourself with those packages as well to improve on the code in future.

christarazi replied:
Ah, good to know. Upon quick inspection of `pkg/eventqueue`, it seems that would actually be a fit for this code. I'll add that as a followup. Thanks

christarazi committed:
bgp: Add k8s package for MetalLB client

This package provides common K8s client code to be used from the
speaker and manager packages.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer reviewed (approved):
LGTM, with the known caveats that we already discussed. :rocket:

seanmwinn commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -696,6 +696,12 @@ data:
 {{- end }}
 
+{{- if and .Values.bgp.enabled (not .Values.bgp.announce.loadbalancerIP) }}
+{{ fail "BGP was enabled, but no announcements were enabled. Please enable one or more announcements." }}
```
Does this imply that the configuration for bgp is required at startup and can't be applied at a later time?

christarazi replied:
Yes, it does imply that. It will be clearer in the docs https://github.com/cilium/cilium/pull/15519.

seanmwinn commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1581,3 +1581,12 @@ clustermesh:
   enabled: false
+
+# -- Configure BGP
+bgp:
```
I put a lot of effort into alphabetizing values, please keep that work in place.

pchaigno replied:
@seanmwinn No alphabetization can resist a horde of developers unless it's tested in CI :wink:

seanmwinn reviewed (changes requested):
Looks good to me.

Lots of effort went into alphabetizing the values. I would appreciate it if we could continue that with any new additions.

pchaigno reviewed (approved):
I reviewed my codeowners only.
